mlir/lib/Dialect/Math/Transforms/LegalizeToF32.cpp

```diff
@@ -109,4 +135,14 @@ void LegalizeToF32Pass::runOnOperation() {
   math::populateLegalizeToF32Patterns(patterns, typeConverter);
   if (failed(applyPartialConversion(op, target, std::move(patterns))))
     return signalPassFailure();
+  
+  if (useCanonicalizeF32Promotion) {
```

Comment:
1. The pass `option` is for users who concerns about the numerical difference. With the option, they can easily switch on / off the optimization.
2. I really appreciate your `one stage` approach, which directly modifies the `matchAndRewrite()` to determine whether to insert `extf / truncf` at the time of the current op is hit. However, if users creates `extf / truncf` explicitly in the IR, then the op pairs cannot be optimized in this way. I think the `two stage` approach can handle such case way more easily.

Reply:
I propose the one stage approach because it *doesn't* optimize explicit truncf / extf pairs 

Explicitly rewriting away all truncf/extf pairs shouldn't be hiding in a type legalization. The legalization can, using the one stage approach, refrain from creating such pairs to improve numerical precision, but it should not eliminate existing ones.

Reply:
Maybe we can apply the one stage approach in legalization pass, and create another pass for something like graph simplification use. @ZhennanQin

Reply:
@krzysz00  May I know what's the difference between the existing truncf / extf and auto-generated ones? Why we can only eliminate the truncf / extf generated from `legalize-to-f32`, but not from any other passes? Would you please provide a use scenario?